File: archive/Facade_ComplicatedSubSystem.cs
// The Facade design pattern.  See Facade_Class.cs for details.
//
// This is a "complicated" sub-system for which a facade will be provided to
// simplify the use of the sub-system.
//
// For this example, there are only three classes.  In a larger sub-system,
// there could be a dozen or more classes and interfaces representing a truly
// dizzying array of functionality.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DesignPatternExamples
{
    /// <summary>
    /// Identifies the type of TAPs that can appear in a TAP chain.
    /// </summary>
    public enum TAPTypes
    {
        /// <summary>
        /// TAP controller.  This is always visible.
        /// </summary>
        CLTAP,
        /// <summary>
        /// Core device
        /// </summary>
        CORE, 
        /// <summary>
        /// GTE device
        /// </summary>
        GTE,
        /// <summary>
        /// PCH device
        /// </summary>
        PCH,
        /// <summary>
        /// PMC device
        /// </summary>
        PMC,
    }


    /// <summary>
    /// Represents a single TAP.
    /// </summary>
    public class TAPNode
    {
        /// <summary>
        /// Whether the TAP is visible in the TAP chain.
        /// </summary>
        public bool Visible;
        
        /// <summary>
        /// Name of this TAP.
        /// </summary>
        public string Name;

        /// <summary>
        /// The idcode for this TAP.
        /// </summary>
        public uint Idcode;

        /// <summary>
        /// A value from the TAPTypes enumeration identifying the type of the
        /// TAP.
        /// </summary>
        public TAPTypes TAPType;

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="name">Name to use.</param>
        /// <param name="idcode">idcode for the TAP.</param>
        /// <param name="tapType">Value from the TAPTypes enumeration.</param>
        /// <param name="initiallyVisible">true if initially visible; otherwise false.</param>
        public TAPNode(string name, uint idcode, TAPTypes tapType, bool initiallyVisible)
        {
            Name = name;
            Idcode = idcode;
            TAPType = tapType;
            Visible = initiallyVisible;
        }
    }


    /// <summary>
    /// Represents a TAP chain, which is a collection of TAPNode objects.
    /// </summary>
    public class TAPChain
    {
        /// <summary>
        /// The list of TAPNodes on this TAP chain.
        /// </summary>
        private List<TAPNode> _nodes = new List<TAPNode>();

        /// <summary>
        /// The Name of this TAP chain.
        /// </summary>
        public string Name;

        /// <summary>
        /// Whether this TAP chain is locked for access.
        /// </summary>
        public bool IsLocked;

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param name="name">Name of this TAP chain.</param>
        public TAPChain(string name)
        {
            Name = name;
        }

        /// <summary>
        /// Helper method to show or hide TAPs on the TAP chain.
        /// </summary>
        /// <param name="nodeSelectMask">A bit mask where the position of each bit
        /// corresponds to a device in the TAP chain, with bit 0 being the first TAP,
        /// bit 1 being the second TAP, and so on.</param>
        /// <param name="makeVisible">true if the TAP is to be made visible on the
        /// TAP chain; otherwise false, the TAP cannot be seen on the TAP chain.</param>
        private void _ShowHideNodes(uint nodeSelectMask, bool makeVisible)
        {
            uint bitMask = 0x2; // bit 0 is always the CLTAP and is always selected
            int numNodes = _nodes.Count;

            // Start at the device after the CLTAP
            for (int index = 1; index < numNodes; ++index)
            {
                if ((bitMask & nodeSelectMask) != 0)
                {
                    _nodes[index].Visible = makeVisible;
                }
                bitMask <<= 1;
                if (bitMask == 0)
                {
                    // We don't allow more than 32 devices
                    break;
                }
            }
        }


        /// <summary>
        /// Helper method to add a TAPNode to the TAP chain.  TAPNode objects
        /// that are of TapTypes.CLTAP are always inserted as the first TAP in
        /// the TAP chain, with the assumption there is only one CLTAP in a
        /// given TAP chain (this is not actually enforced, though).
        /// </summary>
        /// <param name="node">A TAPNode object to add to the tap chain.</param>
        internal void AddNode(TAPNode node)
        {
            if (_nodes.Count > 0 && node.TAPType == TAPTypes.CLTAP)
            {
                // CLTAP always goes at the start of the list.
                _nodes.Insert(0, node);
            }
            else
            {
                _nodes.Add(node);
            }
        }


        /// <summary>
        /// Resets the TAP chain so that all TAPs that are not CLTAPs are no
        /// longer visible in the TAP chain.
        /// </summary>
        public void ResetVisibility()
        {
            foreach (TAPNode node in _nodes)
            {
                if (node.TAPType != TAPTypes.CLTAP)
                {
                    node.Visible = false;
                }
            }
        }

        /// <summary>
        /// Make visible one or more TAPs in the TAP chain.
        /// </summary>
        /// <param name="nodeSelectMask">a bit mask specifying which TAP or TAPs
        /// to make visible, where bit 0 is the first TAP, bit 1 is the second, etc.
        /// Bit 0 is ignored as the first TAP is always visible.</param>
        public void SelectNodes(uint nodeSelectMask)
        {
            _ShowHideNodes(nodeSelectMask, true);
        }

        /// <summary>
        /// Make invisible one or more TAPs in the TAP chain.
        /// </summary>
        /// <param name="nodeSelectMask">a bit mask specifying which TAP or TAPs
        /// to hide, where bit 0 is the first TAP, bit 1 is the second, etc.
        /// Bit 0 is ignored as the first TAP is always visible.</param>
        public void DeselectNodes(uint nodeSelectMask)
        {
            _ShowHideNodes(nodeSelectMask, false);
        }


        /// <summary>
        /// Retrieve a list of idcodes for all TAPs that are visible in
        /// the TAP chain.
        /// </summary>
        /// <returns>Returns an array of uints corresponding to the idcodes of
        /// each visible TAP.  The first idcode corresponds to the first visible
        /// TAP.</returns>
        public uint[] GetIdCodesForVisibleNodes()
        {
            List<uint> idcodes = new List<uint>();

            foreach(TAPNode node in _nodes)
            {
                if (node.Visible)
                {
                    idcodes.Add(node.Idcode);
                }
            }

            return idcodes.ToArray();
        }
    }


    /// <summary>
    /// Represents some kind of system that contains multiple TAP chains.
    /// </summary>
    /// <remarks>
    /// The following operations are available:
    ///  - GetNumChains()
    ///  - LockTAPs()
    ///  - UnlockTAPs()
    ///  - ResetTAPs()
    ///  - SelectTAPs()
    ///  - DeselectTAPs()
    ///  - GetIdcodes()
    /// </remarks>
    public class Facade_ComplicatedSubSystem : ITAPNetwork
    {
        /// <summary>
        /// The list of TAP chains.  In this case, there are two.
        /// </summary>
        private TAPChain[] _tapChains = { new TAPChain("CHAIN0"), new TAPChain("CHAIN1") };

        /// <summary>
        /// Class factory for the sub-system class.
        /// </summary>
        /// <returns>Returns a new instance of the class.</returns>
        public static ITAPNetwork CreateInstance()
        {
            return new Facade_ComplicatedSubSystem();
        }

        /// <summary>
        /// (private) Constructor.  Sets up the TAP chains.
        /// </summary>
        private Facade_ComplicatedSubSystem()
        {
            _tapChains[0].AddNode(new TAPNode("DDD_CLTAPC0", 0x10101010, TAPTypes.CLTAP, true));
            _tapChains[0].AddNode(new TAPNode("DDD_CORE0", 0x20202020, TAPTypes.CORE, false));
            _tapChains[0].AddNode(new TAPNode("DDD_GTE0", 0x30303030, TAPTypes.GTE, false));
            
            _tapChains[1].AddNode(new TAPNode("DDD_CLTAPC1", 0x10101011, TAPTypes.CLTAP, true));
            _tapChains[1].AddNode(new TAPNode("DDD_PCH0", 0x40404040, TAPTypes.PCH, false));
            _tapChains[1].AddNode(new TAPNode("DDD_PMC0", 0x50505050, TAPTypes.PMC, false));
        }

        #region ITapNetwork methods
        /// <summary>
        /// Retrieve the number of TAP chains.
        /// </summary>
        /// <returns>Returns the number of TAP chains</returns>
        public int GetNumChains()
        {
            return _tapChains.Length;
        }

        /// <summary>
        /// Lock the specified TAP chain to indicate exclusive access is
        /// desired.
        /// </summary>
        /// <param name="chainIndex">The index of the TAP chain to access (0..n-1).</param>
        /// <returns>Returns true if the TAP chain was successfully locked; otherwise,
        /// returns false (chain index out of range or the TAP chain is already locked)</returns>
        public bool LockTAPs(int chainIndex)
        {
            bool locked = false;

            if (chainIndex >= 0 && chainIndex < _tapChains.Length)
            {
                if (!_tapChains[chainIndex].IsLocked)
                {
                    _tapChains[chainIndex].IsLocked = true;
                    locked = true;
                }
            }

            return locked;
        }

        /// <summary>
        /// Unlock the specified TAP chain to indicate exclusive access is no
        /// longer desired.
        /// </summary>
        /// <param name="chainIndex">The index of the TAP chain to access (0..n-1).</param>
        /// <returns>Returns true if the TAP chain was successfully unlocked; otherwise,
        /// returns false (chain index out of range or the TAP chain is already unlocked)</returns>
        public bool UnlockTAPs(int chainIndex)
        {
            bool unlocked = false;

            if (chainIndex >= 0 && chainIndex < _tapChains.Length)
            {
                if (_tapChains[chainIndex].IsLocked)
                {
                    _tapChains[chainIndex].IsLocked = false;
                    unlocked = true;
                }
            }

            return unlocked;
        }

        /// <summary>
        /// Reset the visibility of all TAPs on the given TAP chain so that
        /// all TAPs except the first are not visible.
        /// </summary>
        /// <param name="chainIndex">The index of the TAP chain to access (0..n-1).</param>
        public void ResetTAPs(int chainIndex)
        {
            if (chainIndex >= 0 && chainIndex < _tapChains.Length)
            {
                _tapChains[chainIndex].ResetVisibility();
            }
        }

        /// <summary>
        /// Select one or more TAPs in the given TAP chain so those TAPs are
        /// visible.
        /// </summary>
        /// <param name="chainIndex">The index of the TAP chain to access (0..n-1).</param>
        /// <param name="tapSelectMask">A bit mask indicating which TAPs to make
        /// visible, with bit 0 corresponding to the first TAP, bit 1 corresponding
        /// to the second TAP, etc.  Bit 0 is ignored as the first TAP is always
        /// visible.</param>
        public void SelectTAPs(int chainIndex, uint tapSelectMask)
        {
            if (chainIndex >= 0 && chainIndex < _tapChains.Length)
            {
                _tapChains[chainIndex].SelectNodes(tapSelectMask);
            }
        }

        /// <summary>
        /// Deselect one or more TAPs in the given TAP chain so those TAPs are
        /// no longer visible.
        /// </summary>
        /// <param name="chainIndex">The index of the TAP chain to access (0..n-1).</param>
        /// <param name="tapSelectMask">A bit mask indicating which TAPs to make
        /// invisible, with bit 0 corresponding to the first TAP, bit 1 corresponding
        /// to the second TAP, etc.  Bit 0 is ignored as the first TAP is always
        /// visible.</param>
        public void DeselectTAPs(int chainIndex, uint tapSelectMask)
        {
            if (chainIndex >= 0 && chainIndex < _tapChains.Length)
            {
                _tapChains[chainIndex].DeselectNodes(tapSelectMask);
            }
        }

        /// <summary>
        /// Retrieve a list of idcodes for each TAP in the given TAP chain
        /// that is visible.
        /// </summary>
        /// <param name="chainIndex">The index of the TAP chain to access (0..n-1).</param>
        /// <returns>An array of uints holding the idcodes for each TAP, with the
        /// first idcode corresponding to the first visible TAP.</returns>
        public uint[] GetIdcodes(int chainIndex)
        {
            uint[] idcodes = null;

            if (chainIndex >= 0 && chainIndex < _tapChains.Length)
            {
                idcodes = _tapChains[chainIndex].GetIdCodesForVisibleNodes();
            }
            return idcodes;
        }
        #endregion
    }


    /// <summary>
    /// Class factory for the complicated sub-system class.
    /// </summary>
    /// <remarks>
    /// Hides the details of constructing the complicated sub-system and
    /// enforces a singleton nature on the instance of the sub-system.
    /// </remarks>
    public static class Facade_ComplicatedSubSystemFactory
    {
        /// <summary>
        /// A singleton instance of the sub-system.
        /// </summary>
        static ITAPNetwork _instance;


        /// <summary>
        /// Class factory for a singleton instance of the sub-system class.
        /// </summary>
        /// <returns>Returns a singleton instance of the class.</returns>
        public static ITAPNetwork CreateInstance()
        {
            if (_instance == null)
            {
                _instance = Facade_ComplicatedSubSystem.CreateInstance();
            }
            return _instance;
        }
    }
}

File: rust/design_pattern_examples_rust/src/facade/facade_complicatedsubsystem.rs
//! Contains the implementation of the complicated facade sub-system.

use super::facade_idevicenetworklowlevel_trait::IDeviceNetworkLowLevel;

//-----------------------------------------------------------------------------

/// Identifies the type of devices that can appear in a device chain.
/// Part of the "Facade" pattern example.
enum DeviceTypes {
    /// device controller.  This is always visible.
    DEVICECONTROLLER,
    /// Core device
    CORE,
    /// GTE device
    GTE,
    /// PCH device
    PCH,
    /// PMC device
    PMC,
}

//=============================================================================

/// Represents a single device in a device chain.
/// Part of the "Facade" pattern example.
struct DeviceNode {
    /// Whether the device is visible in the device chain.
    visible: bool,

    /// Name of this device.
    _name : String,

    /// The idcode for this device.
    idcode: u32,

    /// A value from the DeviceTypes enumeration identifying the type of the
    /// device.
    device_type : DeviceTypes,
}

impl DeviceNode {
    /// Constructor.
    ///
    /// # Parameters
    /// - name
    ///
    ///   Name to use.
    /// - idcode
    ///
    ///   idcode for the device.
    /// - tap_type
    ///
    ///   Value from the DeviceTypes enumeration.
    /// - initially_visible
    ///
    ///   true if initially visible; otherwise false.
    ///
    /// # Returns
    /// Returns a new DeviceNode instance.
    fn new(name: &str, idcode: u32, tap_type: DeviceTypes, initially_visible: bool) -> DeviceNode {
        DeviceNode {
            _name: name.to_string(),
            idcode: idcode,
            device_type: tap_type,
            visible: initially_visible,
        }
    }
}

//=============================================================================

/// Represents a device chain, which is a collection of DeviceNode objects.
/// Part of the "Facade" pattern example.
struct DeviceChain {
    /// The list of DeviceNodes on this device chain.
    _nodes: Vec<DeviceNode>,

    /// The Name of this device chain.
    _name: String,

    /// Whether this device chain is locked for access.
    locked: bool,
}

impl DeviceChain {
    /// Constructor.
    ///
    /// # Parameters
    /// - name
    ///
    ///   Name of this device chain
    ///
    /// # Returns
    /// Returns a new DeviceChain instance.
    pub fn new(name: &str) -> DeviceChain {
        DeviceChain {
            _name: name.to_string(),
            locked: false,
            _nodes: vec![]
        }
    }

    /// Helper method to add a DeviceNode to the device chain.  DeviceNode
    /// objects that are of DEVICECONTROLLER are always inserted as the first
    /// device in the device chain, with the assumption there is only one
    /// DEVICECONTROLLER in a given device chain (this is not actually
    /// enforced, though).
    ///
    /// # Parameters
    /// - node
    ///
    ///   A DeviceNode object to add to the device chain.
    pub fn add_node(&mut self, node: DeviceNode) {
        match node.device_type {
            DeviceTypes::DEVICECONTROLLER => self._nodes.insert(0, node),
            _ => self._nodes.push(node)
        }
    }

    /// Resets the device chain so that all devices that are not CLdevices are
    /// no longer visible in the device chain.
    pub fn reset_visibility(&mut self) {
        for node in self._nodes.iter_mut() {
            if let DeviceTypes::DEVICECONTROLLER = node.device_type {
                // Device Controller's visibility is never touched
            } else {
                node.visible = false;
            }
        }
    }

    /// Make visible one or more devices in the device chain.
    ///
    /// # Parameters
    /// - node_select_mask
    ///
    ///   A bit mask specifying which device or devices to make visible, where
    ///   bit 0 is the first device, bit 1 is the second, etc.  Bit 0 is
    ///   ignored as the first device is always visible.
    pub fn select_nodes(&mut self, node_select_mask: u32) {
        self._show_hide_nodes(node_select_mask, true);

    }

    /// Make invisible one or more devices in the device chain.
    ///
    ///
    /// # Parameters
    /// - node_select_mask
    ///
    /// A bit mask specifying which device or devices to hide, where bit 0 is
    /// the first device, bit 1 is the second, etc.  Bit 0 is ignored as the
    ///  first device is always visible.
    pub fn deselect_nodes(&mut self, node_select_mask: u32) {
        self._show_hide_nodes(node_select_mask, false);
    }
   
    /// Retrieve a list of idcodes for all devices that are visible in
    /// the device chain.
    ///
    /// # Returns
    /// Returns an array of unsigned 32-bit integers corresponding to the
    /// idcodes of each visible device.  The first idcode corresponds to the
    /// first visible device.
    pub fn get_id_codes_for_visible_nodes(&self) -> Vec<u32> {
        let mut idcodes : Vec<u32> = vec![];

        for node in self._nodes.iter() {
            if node.visible {
                idcodes.push(node.idcode);
            }
        }
        idcodes
    }


    /// Helper method to show or hide devices on the device chain.
    ///
    /// # Parameters
    /// - node_select_mask
    ///
    ///   A bit mask where the position of each bit corresponds to a device in
    ///   the device chain, with bit 0 being the first device, bit 1 being the
    ///   second device, and so on.
    /// - make_visible
    ///
    ///   true if the device is to be made visible on the device chain;
    ///   otherwise false, the device cannot be seen on the device chain.
    fn _show_hide_nodes(&mut self, node_select_mask: u32, make_visible: bool) {
        let mut bit_mask: u32 = 0x2; // bit 0 is always the DEVICECONTROLLER and is always selected
        let num_nodes = self._nodes.len();

        // Start at the device after the DEVICECONTROLLER
        for index in 1..num_nodes {
            if (bit_mask & node_select_mask) != 0 {
                self._nodes[index].visible = make_visible;
            }
            bit_mask <<= 1;
            if bit_mask == 0 {
                // We don't allow more than 32 devices
                break;
            }
        }
    }
}

//-----------------------------------------------------------------------------

/// Represents some kind of system that contains multiple device chains.
/// Part of the "Facade" pattern example.
///
/// The following operations are available:
///  - get_num_chains()
///  - lock_devices()
///  - unlock_devices()
///  - reset_devices()
///  - select_devices()
///  - deselect_devices()
///  - get_idcodes()
pub struct FacadeComplicatedSubSystem {
    device_chains: Vec<DeviceChain>,
}

impl FacadeComplicatedSubSystem {
    /// Constructor.
    pub fn new() -> Box<dyn IDeviceNetworkLowLevel> {
        let mut device_chain0 = DeviceChain::new("CHAIN0");
        let mut device_chain1 = DeviceChain::new("CHAIN1");

        device_chain0.add_node(DeviceNode::new("DDD_DEVCTRL0", 0x10101010, DeviceTypes::DEVICECONTROLLER, true));
        device_chain0.add_node(DeviceNode::new("DDD_CORE0", 0x20202020, DeviceTypes::CORE, false));
        device_chain0.add_node(DeviceNode::new("DDD_GTE0", 0x30303030, DeviceTypes::GTE, false));

        device_chain1.add_node(DeviceNode::new("DDD_DEVCTRL1", 0x10101011, DeviceTypes::DEVICECONTROLLER, true));
        device_chain1.add_node(DeviceNode::new("DDD_PCH0", 0x40404040, DeviceTypes::PCH, false));
        device_chain1.add_node(DeviceNode::new("DDD_PMC0", 0x50505050, DeviceTypes::PMC, false));

        Box::new(FacadeComplicatedSubSystem {
            device_chains: vec![device_chain0, device_chain1],
        })
    }
}

impl IDeviceNetworkLowLevel for FacadeComplicatedSubSystem {
    fn get_num_chains(&self) -> usize {
        self.device_chains.len()
    }

    fn lock_device_chain(&mut self, chain_index: usize) -> bool {
        let mut locked = false;
        if chain_index < self.device_chains.len() {
            if !self.device_chains[chain_index].locked {
                self.device_chains[chain_index].locked = true;
                locked = true;
            }
        }
        locked
    }

    fn unlock_device_chain(&mut self, chain_index: usize) -> bool {
        let mut unlocked = false;
        if chain_index < self.device_chains.len() {
            if self.device_chains[chain_index].locked {
                self.device_chains[chain_index].locked = false;
                unlocked = true;
            }
        }
        unlocked
    }

    fn reset_device_chain(&mut self, chain_index: usize) {
        if chain_index < self.device_chains.len() {
            self.device_chains[chain_index].reset_visibility();
        }
    }

    fn enable_devices_in_device_chain(&mut self, chain_index: usize, devices_select_mask: u32) {
        if chain_index < self.device_chains.len() {
            self.device_chains[chain_index].select_nodes(devices_select_mask);
        }
    }

    fn disable_devices_in_device_chain(&mut self, chain_index: usize, devices_select_mask: u32) {
        if chain_index < self.device_chains.len() {
            self.device_chains[chain_index].deselect_nodes(devices_select_mask);
        }
    }

    fn get_idcodes(&self, chain_index: usize) -> Vec<u32> {
        let mut idcodes: Vec<u32> = vec![];

        if chain_index < self.device_chains.len() {
            idcodes = self.device_chains[chain_index].get_id_codes_for_visible_nodes();
        }

        idcodes
    }
}